File: ui/packages/renderer/src/bindings/api/workspace/workspace.ts
import { invokeAsync } from '@/bindings/core';

import { FileTree, FileTreeSortType, Workspace, WorkspaceSettings } from './types';
import { getCurrentOpenWorkspace } from './workspaceManager';

const checkCurrentOpenWorkspace = async (): Promise<string> => {
  const path = getCurrentOpenWorkspace();
  if (!path) throw new Error('workspace is not open');
  return path;
};

export const workspace = {
  isOpenWorkspace: async (path: string): Promise<boolean> => {
    return path ? (await invokeAsync('workspace.is_open_workspace', { path }))! : false;
  },
  getWorkspace: async (path: string): Promise<Workspace | null> => {
    return path ? (await invokeAsync('workspace.get_open_workspace', { path }))! : null;
  },
  getWorkspaceSettings: async (path: string): Promise<WorkspaceSettings> => {
    return (await invokeAsync('workspace.get_open_workspace_settings', { path }))!;
  },
  setWorkspaceSettings: async (path: string, settings: WorkspaceSettings): Promise<Workspace> => {
    return (await invokeAsync('workspace.set_open_workspace_settings', { path, settings }))!;
  },
  getOpenWorkspaceFileTree: async (path: string): Promise<FileTree> => {
    return (await invokeAsync('workspace.get_open_workspace_file_tree', { path }))!;
  },
  setOpenWorkspaceFileTreeSortType: async (
    path: string,
    sortType: FileTreeSortType,
  ): Promise<void> => {
    return (await invokeAsync('workspace.set_open_workspace_file_tree_sort_type', {
      path,
      sortType,
    }))!;
  },
  setOpenWorkspaceFollowGitignore: async (
    path: string,
    followGitignore: boolean,
  ): Promise<void> => {
    return (await invokeAsync('workspace.set_open_workspace_follow_gitignore', {
      path,
      followGitignore,
    }))!;
  },
  setOpenWorkspaceCustomIgnore: async (path: string, customIgnore: string): Promise<void> => {
    return (await invokeAsync('workspace.set_open_workspace_custom_ignore', {
      path,
      customIgnore,
    }))!;
  },
  resetOpenWorkspaceCustomIgnore: async (path: string): Promise<void> => {
    return (await invokeAsync('workspace.reset_open_workspace_custom_ignore', {
      path,
    }))!;
  },
  resetOpenWorkspaceHistroySnapshootCount: async (path: string): Promise<void> => {
    return (await invokeAsync('workspace.reset_open_workspace_histroy_snapshoot_count', {
      path,
    }))!;
  },
  resetOpenWorkspaceUploadAttachmentPath: async (path: string): Promise<void> => {
    return (await invokeAsync('workspace.reset_open_workspace_upload_attachment_path', {
      path,
    }))!;
  },
  resetOpenWorkspaceUploadImagePath: async (path: string): Promise<void> => {
    return (await invokeAsync('workspace.reset_open_workspace_upload_image_path', {
      path,
    }))!;
  },
  callOpenWorkspaceReinit: async (path: string): Promise<void> => {
    return (await invokeAsync('workspace.call_open_workspace_reinit', { path }))!;
  },
  getCurrentWorkspace: async (): Promise<Workspace | null> => {
    const path = getCurrentOpenWorkspace();
    return path ? await workspace.getWorkspace(path) : null;
  },
  getCurrentWorkspaceSettings: async (): Promise<WorkspaceSettings> => {
    const path = await checkCurrentOpenWorkspace();
    return await workspace.getWorkspaceSettings(path);
  },
  setCurrentWorkspaceSettings: async (settings: WorkspaceSettings): Promise<Workspace> => {
    const path = await checkCurrentOpenWorkspace();
    return await workspace.setWorkspaceSettings(path, settings);
  },
  getCurrentworkspaceFileTree: async (): Promise<FileTree> => {
    const path = await checkCurrentOpenWorkspace();
    return await workspace.getOpenWorkspaceFileTree(path);
  },
  setCurrentWorkspaceFileTreeSortType: async (sortType: FileTreeSortType): Promise<void> => {
    const path = await checkCurrentOpenWorkspace();
    return await workspace.setOpenWorkspaceFileTreeSortType(path, sortType);
  },
  setCurrentWorkspaceFollowGitignore: async (followGitignore: boolean): Promise<void> => {
    const path = await checkCurrentOpenWorkspace();
    return await workspace.setOpenWorkspaceFollowGitignore(path, followGitignore);
  },
  setCurrentWorkspaceCustomIgnore: async (customIgnore: string): Promise<void> => {
    const path = await checkCurrentOpenWorkspace();
    return await workspace.setOpenWorkspaceCustomIgnore(path, customIgnore);
  },
  resetCurrentWorkspaceCustomIgnore: async (): Promise<void> => {
    const path = await checkCurrentOpenWorkspace();
    return await workspace.resetOpenWorkspaceCustomIgnore(path);
  },
  resetCurrentWorkspaceHistroySnapshootCount: async (): Promise<void> => {
    const path = await checkCurrentOpenWorkspace();
    return await workspace.resetOpenWorkspaceHistroySnapshootCount(path);
  },
  resetCurrentWorkspaceUploadAttachmentPath: async (): Promise<void> => {
    const path = await checkCurrentOpenWorkspace();
    return await workspace.resetOpenWorkspaceUploadAttachmentPath(path);
  },
  resetCurrentWorkspaceUploadImagePath: async (): Promise<void> => {
    const path = await checkCurrentOpenWorkspace();
    return await workspace.resetOpenWorkspaceUploadImagePath(path);
  },
  reinitCurrentworkspace: async (): Promise<void> => {
    const path = await checkCurrentOpenWorkspace();
    await workspace.callOpenWorkspaceReinit(path);
  },
};

File: ui/flutter/lib/src/bindings/api/workspace/workspace.dart
import 'package:lonanote/src/bindings/api/workspace/types.dart';
import 'package:lonanote/src/bindings/api/workspace/workspace_manager.dart';
import 'package:lonanote/src/common/config/app_config.dart';

import '../../bindings.dart';

String checkCurrentOpenWorkspace() {
  final path = RustWorkspaceManager.getCurrentOpenWorkspace();
  if (path == null) throw Exception('workspace is not open');
  return path;
}

class RustWorkspace {
  static Future<bool> isOpenWorkspace(String? path) async {
    if (path == null) {
      return false;
    }
    return await Bindings.invokeAsync(
        key: "workspace.is_open_workspace", value: {"path": path});
  }

  static Future<RustWorkspaceData?> getWorkspace(String? path) async {
    if (path == null) {
      return null;
    }
    final ws = await Bindings.invokeAsync(
        key: "workspace.get_open_workspace", value: {"path": path});
    return RustWorkspaceData.fromJson(ws);
  }

  static Future<RustWorkspaceSettings> getWorkspaceSettings(String path) async {
    final s = await Bindings.invokeAsync(
        key: "workspace.get_open_workspace_settings", value: {"path": path});
    return RustWorkspaceSettings.fromJson(s);
  }

  static Future<RustWorkspaceData> setWorkspaceSettings(
      String path, RustWorkspaceSettings settings) async {
    final ws = await Bindings.invokeAsync(
        key: "workspace.set_open_workspace_settings",
        value: {"path": path, "settings": settings.toJson()});
    return RustWorkspaceData.fromJson(ws);
  }

  static Future<RustFileTree> getOpenWorkspaceFileTree(String path) async {
    final f = await Bindings.invokeAsync(
        key: "workspace.get_open_workspace_file_tree", value: {"path": path});
    return RustFileTree.fromJson(f);
  }

  static Future<void> setOpenWorkspaceFileTreeSortType(
      String path, String sortType) async {
    await Bindings.invokeAsync(
        key: "workspace.set_open_workspace_file_tree_sort_type",
        value: {"path": path, "sortType": sortType});
  }

  static Future<void> setOpenWorkspaceFollowGitignore(
      String path, bool followGitignore) async {
    await Bindings.invokeAsync(
        key: "workspace.set_open_workspace_follow_gitignore",
        value: {"path": path, "followGitignore": followGitignore});
  }

  static Future<void> setOpenWorkspaceCustomIgnore(
      String path, String customIgnore) async {
    await Bindings.invokeAsync(
        key: "workspace.set_open_workspace_custom_ignore",
        value: {"path": path, "customIgnore": customIgnore});
  }

  static Future<void> resetOpenWorkspaceCustomIgnore(String path) async {
    await Bindings.invokeAsync(
        key: "workspace.reset_open_workspace_custom_ignore",
        value: {"path": path});
  }

  static Future<void> resetOpenWorkspaceHistroySnapshootCount(
      String path) async {
    await Bindings.invokeAsync(
        key: "workspace.reset_open_workspace_histroy_snapshoot_count",
        value: {"path": path});
  }

  static Future<void> resetOpenWorkspaceUploadAttachmentPath(
      String path) async {
    await Bindings.invokeAsync(
        key: "workspace.reset_open_workspace_upload_attachment_path",
        value: {"path": path});
  }

  static Future<void> resetOpenWorkspaceUploadImagePath(String path) async {
    await Bindings.invokeAsync(
        key: "workspace.reset_open_workspace_upload_image_path",
        value: {"path": path});
  }

  static Future<void> callOpenWorkspaceReinit(String path) async {
    await Bindings.invokeAsync(
        key: "workspace.call_open_workspace_reinit", value: {"path": path});
  }

  static Future<RustWorkspaceData?> getCurrentWorkspace() async {
    final path = RustWorkspaceManager.getCurrentOpenWorkspace();
    return await getWorkspace(path);
  }

  static Future<RustWorkspaceSettings?> getCurrentWorkspaceSettings() async {
    final path = checkCurrentOpenWorkspace();
    return await getWorkspaceSettings(path);
  }

  static Future<RustWorkspaceData> setCurrentWorkspaceSettings(
      RustWorkspaceSettings settings) async {
    final path = checkCurrentOpenWorkspace();
    return await setWorkspaceSettings(path, settings);
  }

  static Future<RustFileTree> getCurrentworkspaceFileTree() async {
    final path = checkCurrentOpenWorkspace();
    return await getOpenWorkspaceFileTree(path);
  }

  static Future<void> setCurrentWorkspaceFileTreeSortType(
      String sortType) async {
    final path = checkCurrentOpenWorkspace();
    await setOpenWorkspaceFileTreeSortType(path, sortType);
  }

  static Future<void> setCurrentWorkspaceFollowGitignore(
      bool followGitignore) async {
    final path = checkCurrentOpenWorkspace();
    await setOpenWorkspaceFollowGitignore(path, followGitignore);
  }

  static Future<void> setCurrentWorkspaceCustomIgnore(
      String customIgnore) async {
    final path = checkCurrentOpenWorkspace();
    await setOpenWorkspaceCustomIgnore(path, customIgnore);
  }

  static Future<void> resetCurrentWorkspaceCustomIgnore() async {
    final path = checkCurrentOpenWorkspace();
    await resetOpenWorkspaceCustomIgnore(path);
  }

  static Future<void> resetCurrentWorkspaceHistroySnapshootCount() async {
    final path = checkCurrentOpenWorkspace();
    await resetOpenWorkspaceHistroySnapshootCount(path);
  }

  static Future<void> resetCurrentWorkspaceUploadAttachmentPath() async {
    final path = checkCurrentOpenWorkspace();
    await resetOpenWorkspaceUploadAttachmentPath(path);
  }

  static Future<void> resetCurrentWorkspaceUploadImagePath() async {
    final path = checkCurrentOpenWorkspace();
    await resetOpenWorkspaceUploadImagePath(path);
  }

  static Future<void> reinitCurrentworkspace() async {
    final path = checkCurrentOpenWorkspace();
    await callOpenWorkspaceReinit(path);
  }

  static Future<void> test() async {
    if (!AppConfig.isDebug) return;
  }
}
